### PR TITLE
Avoid `B015`,`B018` for last expression in a cell

### PR DESCRIPTION
**Pull request by dhruvmanila**
## Summary

This PR updates `B015` and `B018` to ignore last top-level expressions in each
cell of a Jupyter Notebook.

Part of #8669

## Test Plan

Add test cases for both rules and update the snapshots.

### CODE REVIEW

**Comment by dhruvmanila** [comment]
Current dependencies on/for this PR:
* main
  * **PR #8813** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8813?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #8814** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8814?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #8815** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/8815?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/8815?utm_source=stack-comment).

File: crates/ruff_linter/src/rules/flake8_bugbear/helpers.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    cell_offsets: Option<&CellOffsets>,
+) -> bool {
+    if !semantic.at_top_level() {
+        return false;
+    }
+    let current_statement_end = semantic.current_statement().end();
+    cell_offsets
+        .and_then(|cell_offsets| cell_offsets.containing_range(current_statement_end))
+        .is_some_and(|cell_range| {
+            !SimpleTokenizer::new(
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/flake8_bugbear/helpers.rs:22]: Maybe a bit easier to understand as `.all(|token| token.kind() == SimpleTokenKind::Semi || token.kind().is_trivia())`? Right now, it's a double negative with `!any` and the internal `!=`.

File: crates/ruff_linter/src/rules/flake8_bugbear/helpers.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+use ruff_notebook::CellOffsets;
+use ruff_python_semantic::SemanticModel;
+use ruff_python_trivia::{SimpleTokenKind, SimpleTokenizer};
+use ruff_source_file::Locator;
+use ruff_text_size::{Ranged, TextRange};
+
+/// Return `true` if the statement containing the current expression is the last
+/// top-level expression in the cell. This assumes that the source is a Jupyter
+/// Notebook.
+pub(crate) fn at_last_top_level_expression_in_cell(
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/flake8_bugbear/helpers.rs:10]: Can probably be `pub(super)`?

**Review by charliermarsh — APPROVED** [review]
Great!

File: crates/ruff_linter/src/rules/flake8_bugbear/helpers.rs (hunk 1 above)
Comment (dhruvmanila) [reply crates/ruff_linter/src/rules/flake8_bugbear/helpers.rs:22]: Thanks! My initial implementation was using `skip_while` which clippy didn't like. This is much better.

**Pull request merged**